includes/UserMailer.php

```diff
@@ -434,95 +412,154 @@ protected static function quotedPrintableCallback( $matches ) {
-					'wl_namespace' => $title->getNamespace(),
-					'wl_user != ' . intval( $editor->getID() ),
-					$notificationTimeoutSql,
-				), __METHOD__
-			);
-			foreach ( $res as $row ) {
-				$watchers[] = intval( $row->wl_user );
-			}
+			$notificationTimeoutSql = $this->getTimeOutSql();
+			$watchers = $this->getWatchersToNotify( $notificationTimeoutSql );
 			if ( $watchers ) {
-				// Update wl_notificationtimestamp for all watching users except
-				// the editor
-				$wl = WatchedItem::fromUserTitle( $editor, $title );
-				$wl->updateWatch( $watchers, $timestamp );
+				$this->updateWatchedItem( $watchers );
 			}
+			wfRunHooks( 'NotifyOnSubPageChange', array ( $watchers, $this->title, $this->editor, $notificationTimeoutSql ) );
+		}
+		if ( $this->shouldSendEmail( $watchers ) ) {
+			$this->actuallyNotifyOnPageChange( $watchers );
+		}
+	}
+
+	/**
+	 * Add a timeout to the watchlist email block
+	 */
+	private function getTimeOutSql(){
+		global $wgEnableWatchlistNotificationTimeout, $wgWatchlistNotificationTimeout;
+
+		if ( !empty( $this->otherParam['notisnull'] ) ) {
+			$notificationTimeoutSql = "1";
+		} elseif ( !empty( $wgEnableWatchlistNotificationTimeout ) && isset( $WatchlistNotificationTimeout ) ) {
+			$blockTimeout = wfTimestamp( TS_MW, wfTimestamp( TS_UNIX, $this->timestamp ) - intval( $wgWatchlistNotificationTimeout ) );
+			$notificationTimeoutSql = "wl_notificationtimestamp IS NULL OR wl_notificationtimestamp < '$blockTimeout'";
+		} else {
+			$notificationTimeoutSql = 'wl_notificationtimestamp IS NULL';
+		}
 
-			/* Wikia change begin - @author: Jakub Kurcek */
-			wfRunHooks( 'NotifyOnSubPageChange', array ( $watchers, $title, $editor, $notificationTimeoutSql ) );
-			/* Wikia change end */
+		return $notificationTimeoutSql;
+	}
+
+	private function getWatchersToNotify( $notificationTimeoutSql ) {
+		$watchers = [];
+		$dbw = wfGetDB( DB_MASTER );
+		$res = $dbw->select( [ 'watchlist' ],
+			[ 'wl_user' ],
+			[
+				'wl_title' => $this->title->getDBkey(),
+				'wl_namespace' => $this->title->getNamespace(),
+				'wl_user != ' . intval( $this->editor->getID() ),
+				$notificationTimeoutSql
+			], __METHOD__
+		);
+
+		foreach ( $res as $row ) {
+			$watchers[] = $row->wl_user;
 		}
 
+		return $watchers;
+	}
+
+	/**
+	 * Update wl_notificationtimestamp for all watching users except the editor
+	 */
+	private function updateWatchedItem( $watchers ) {
+		$wl = WatchedItem::fromUserTitle( $this->editor, $this->title );
+		$wl->updateWatch( $watchers, $this->timestamp );
+	}
+
+	/**
+	 * Check if there are either 1.) users watching the page, or 2.) the page which was
+	 * edited is a User Talk page and the owner of that page wants to receive notifications
+	 * when their User Talk page is changed.
+	 * @param array $watchers
+	 * @return bool
+	 */
+	private function shouldSendEmail( array $watchers ) {
+		global $wgUsersNotifiedOnAllChanges, $wgEnotifMinorEdits, $wgEnotifUserTalk;
+
 		$sendEmail = true;
```

Comment:
This method is unnecessarily complicated. It can be simplified to the following:
```php
		global $wgUsersNotifiedOnAllChanges, $wgEnotifMinorEdits, $wgEnotifUserTalk;

		// If nobody is watching the page, and there are no users notified on all changes
		// don't bother creating a job/trying to send emails
		// $watchers deals with $wgEnotifWatchlist
		return ( count( $watchers ) && count( $wgUsersNotifiedOnAllChanges ) ) &&
			// Only send notification for non minor edits, unless $wgEnotifMinorEdits
			( !$this->minorEdit || ( $wgEnotifMinorEdits && !$this->editor->isAllowed( 'nominornewtalk' ) ) ) &&
			( $wgEnotifUserTalk && $this->isUserTalkPage() && $this->canSendUserTalkEmail() );
```
Also note, as the function stands now, the first conditional smells wrong; I suspect, `||` is more appropriate, as in
`if ( !count( $watchers ) || !count( $wgUsersNotifiedOnAllChanges ) ) {`